File: apps/ios/App/Screens/CompareLapsScreen.swift
import SwiftUI
import TrackEvolutionKit

/// Compare any two laps with telemetry at one track — different sessions, events,
/// even years apart (#165). `viewLapCompare` in `public/app.js` is the reference:
/// the same "current me vs best me" default picks, head-to-head numbers and
/// length-mismatch warning, with the maths in the Kit's `CompareLaps`, pinned to
/// the web implementation by `contracts/logic/compare-laps.json`.
///
/// Presented as a sheet from the track screen, like the event page's channel
/// graphs and for the same reason: the stacked charts want the full height.
struct CompareLapsScreen: View {
    let trackId: Int

    @Environment(AuthController.self) private var auth
    @State private var model: CompareLapsModel?

    private static let kphToMph = 0.621371

    var body: some View {
        TELoadable(state: model?.state ?? .loading, retry: { await model?.load() }) {
            if let model {
                content(model)
            }
        }
        .background(Color(.bgPage))
        .task {
            if model == nil {
                let model = CompareLapsModel(api: auth.api, trackId: trackId)
                self.model = model
                await model.load()
            }
        }
    }

    @ViewBuilder
    private func content(_ model: CompareLapsModel) -> some View {
        if let pair = model.pair {
            ScrollView {
                VStack(alignment: .leading, spacing: 14) {
                    Text("Compare two laps")
                        .teStyle(.h1)
                        .foregroundStyle(Color(.textStrong))
                    picker("Lap A", selection: model.selA, color: Color(.chartLine), model) { model.pick(a: $0) }
                    picker("Lap B", selection: model.selB, color: Color(.chartLineB), model) { model.pick(b: $0) }
                    if pair.mismatch > CompareLaps.LENGTH_MISMATCH_WARN {
                        Text("⚠️ These laps cover driven distances \(Int((pair.mismatch * 100).rounded()))% apart — likely a different layout or start/finish line, so the distance alignment may be off.")
                            .teStyle(.xs)
                            .foregroundStyle(Color(.textMuted))
                    }
                    headToHead(pair)
                    Text("The delta chart shows where time is gained or lost vs the faster lap; the channels below show why. Tap a chart to read values.")
                        .teStyle(.xs)
                        .foregroundStyle(Color(.textFaint))
                    LapChannelPanel(channels: pair.aligned, laps: pair.laps, preselect: [0, 1])
                        // Recreate the panel when the picks change: its highlight
                        // selection is @State seeded on appear, and a stale selection
                        // against new channel data would highlight the wrong laps.
                        .id("\(model.selA)-\(model.selB)")
                }
                .padding(TESpacing.pageGutter)
                .frame(maxWidth: .infinity, alignment: .leading)
            }
        } else {
            TEPage {
                Text("Compare two laps")
                    .teStyle(.h1)
                    .foregroundStyle(Color(.textStrong))
                TEEmpty("Comparing laps needs two laps with telemetry at this track — import a session (or record laps) first.")
            }
        }
    }

    // MARK: - Lap pickers

    private func picker(
        _ label: String, selection: Int, color: Color, _ model: CompareLapsModel, set: @escaping (Int) -> Void
    ) -> some View {
        HStack(spacing: 8) {
            Circle()
                .fill(color)
                .frame(width: 10, height: 10)
            Menu {
                ForEach(model.rows.indices, id: \.self) { index in
                    Button {
                        set(index)
                    } label: {
                        if index == selection {
                            Label(model.pickLabel(index), systemImage: "checkmark")
                        } else {
                            Text(model.pickLabel(index))
                        }
                    }
                }
            } label: {
                HStack(spacing: 6) {
                    Text(model.pickLabel(selection))
                        .teStyle(.bodyStrong)
                        .foregroundStyle(Color(.textStrong))
                        .lineLimit(1)
                        .minimumScaleFactor(0.8)
                    Image(systemName: "chevron.up.chevron.down")
                        .font(.system(size: 11, weight: .semibold))
                        .foregroundStyle(Color(.textFaint))
                }
                .padding(.horizontal, 12)
                .padding(.vertical, 8)
                .background(Color(.surfaceRaised), in: .rect(cornerRadius: TERadius.sm))
                .overlay(
                    RoundedRectangle(cornerRadius: TERadius.sm)
                        .strokeBorder(Color(.borderHairline), lineWidth: 1)
                )
            }
            .accessibilityLabel("\(label): \(model.pickLabel(selection))")
        }
    }

    // MARK: - Head to head

    private func headToHead(_ pair: CompareLapsModel.Pair) -> some View {
        TECard {
            VStack(spacing: 0) {
                statRow("Lap time", fmt: { LapTime.fmtMs($0.timeMs) }, delta: deltaMs(pair), pair)
                divider()
                metric("Top speed", pair, \.topSpeedKph, mph, deltaFmt: mphDelta)
                divider()
                metric("Min speed", pair, \.minSpeedKph, mph, deltaFmt: mphDelta)
                divider()
                metric("Avg speed", pair, \.avgSpeedKph, mph, deltaFmt: mphDelta)
                divider()
                metric("Max RPM", pair, \.maxRpm, { "\(Int($0.rounded()))" }, deltaFmt: { signed($0, "\(Int(abs($0).rounded()))") })
                divider()
                metric("Max lateral G", pair, \.maxLatG, { String(format: "%.2f", $0) }, deltaFmt: { signed($0, String(format: "%.2f", abs($0))) })
                divider()
                metric("Full throttle", pair, \.fullThrottlePct, { "\(Int($0.rounded()))% of lap" }, deltaFmt: ppDelta)
                divider()
                metric("On the brakes", pair, \.brakingPct, { "\(Int($0.rounded()))% of lap" }, deltaFmt: ppDelta)
            }
        }
    }

    private func divider() -> some View {
        Divider().overlay(Color(.borderHairline))
    }

    private func statRow(
        _ label: String, fmt: (CompareLaps.Metrics) -> String, delta: String?, _ pair: CompareLapsModel.Pair
    ) -> some View {
        HStack(spacing: 8) {
            Text(label)
                .teStyle(.sm)
                .foregroundStyle(Color(.textMuted))
            Spacer(minLength: 8)
            Text(fmt(pair.metricsA))
                .teStyle(.lapTime)
                .foregroundStyle(Color(.textStrong))
            Text(fmt(pair.metricsB))
                .teStyle(.lapTime)
                .foregroundStyle(Color(.textStrong))
            Text(delta ?? "—")
                .teStyle(.lapTime)
                .foregroundStyle(Color(.textMuted))
                .frame(minWidth: 56, alignment: .trailing)
        }
        .padding(.vertical, 8)
        .accessibilityElement(children: .ignore)
        .accessibilityLabel("\(label): lap A \(fmt(pair.metricsA)), lap B \(fmt(pair.metricsB)), delta \(delta ?? "none")")
    }

    private func metric(
        _ label: String,
        _ pair: CompareLapsModel.Pair,
        _ key: KeyPath<CompareLaps.Metrics, Double?>,
        _ fmt: (Double) -> String,
        deltaFmt: (Double) -> String
    ) -> some View {
        let a = pair.metricsA[keyPath: key]
        let b = pair.metricsB[keyPath: key]
        let delta = (a != nil && b != nil) ? deltaFmt(b! - a!) : nil
        return statRow(
            label,
            fmt: { $0[keyPath: key].map(fmt) ?? "—" },
            delta: delta,
            pair
        )
    }

    private func deltaMs(_ pair: CompareLapsModel.Pair) -> String {
        LapTime.fmtDelta(pair.metricsB.timeMs - pair.metricsA.timeMs)
    }

    private func mph(_ kph: Double) -> String {
        "\(Int((kph * Self.kphToMph).rounded())) mph"
    }

    private func mphDelta(_ kph: Double) -> String {
        signed(kph, "\(Int((abs(kph) * Self.kphToMph).rounded())) mph")
    }

    private func ppDelta(_ d: Double) -> String {
        signed(d, String(format: "%.1fpp", abs(d)))
    }

    /// The sign is the message, so it is always shown.
    private func signed(_ value: Double, _ magnitude: String) -> String {
        value < 0 ? "−\(magnitude)" : "+\(magnitude)"
    }
}

/// The compare screen's data: every event detail at the track with channel data,
/// flattened to pickable laps, and the currently selected pair.
@MainActor
@Observable
final class CompareLapsModel {
    private let api: APIClient
    let trackId: Int

    private(set) var state: LoadState = .loading
    private(set) var rows: [CompareLaps.Row] = []
    /// `sessions.channels` by session id, so a picked row finds its entry.
    private var channelsBySession: [Int: SessionChannels] = [:]

    /// Indexes into `rows` — seeded by `CompareLaps.defaultComparePicks`.
    var selA = 0
    var selB = 0

    init(api: APIClient, trackId: Int) {
        self.api = api
        self.trackId = trackId
    }

    func load() async {
        do {
            let events = try await api.events(trackId: trackId)
            var details: [EventDetail] = []
            for event in events where event.lapCount > 0 {
                details.append(try await api.event(id: event.id))
            }
            rows = CompareLaps.comparableLaps(details.map(CompareLaps.EventLaps.init(detail:)))
            channelsBySession = details.reduce(into: [:]) { acc, detail in
                for session in detail.sessions {
                    acc[session.id] = session.channels
                }
            }
            if let picks = CompareLaps.defaultComparePicks(rows) {
                selA = picks.a
                selB = picks.b
            }
            state = .ready
        } catch let error as APIError {
            state = .failed(error.message)
        } catch {
            state = .failed(error.localizedDescription)
        }
    }

    /// Picking the lap the other side already shows swaps the two instead of
    /// comparing a lap to itself — the web view's rule.
    func pick(a index: Int) {
        if index == selB { selB = selA }
        selA = index
    }

    func pick(b index: Int) {
        if index == selA { selA = selB }
        selB = index
    }

    /// "May 3, 2026 · Sat AM · Lap 2 — 1:30.480", trimmed to what the row has.
    func pickLabel(_ index: Int) -> String {
        guard rows.indices.contains(index) else { return "—" }
        let row = rows[index]
        let parts = [EventDates.fmtDate(row.date), row.sessionLabel, "Lap \(String(row.lapNum))"]
        return "\(parts.compactMap(\.self).joined(separator: " · ")) — \(LapTime.fmtMs(row.timeMs))"
    }

    /// Everything the view needs for the current pair, or nil when fewer than two
    /// laps have channel data.
    struct Pair {
        var aligned: SessionChannels
        /// Synthetic lap rows for the panel's chips: ids 0 and 1, the rows' own lap
        /// numbers, and the entries' times so `matchLapsToChannels` pairs them.
        var laps: [Lap]
        var metricsA: CompareLaps.Metrics
        var metricsB: CompareLaps.Metrics
        var mismatch: Double
    }

    var pair: Pair? {
        guard rows.count >= 2,
              rows.indices.contains(selA), rows.indices.contains(selB), selA != selB,
              let chanA = channelsBySession[rows[selA].sessionId],
              let chanB = channelsBySession[rows[selB].sessionId],
              chanA.laps.indices.contains(rows[selA].chIdx),
              chanB.laps.indices.contains(rows[selB].chIdx)
        else { return nil }
        let entryA = chanA.laps[rows[selA].chIdx]
        let entryB = chanB.laps[rows[selB].chIdx]
        let aligned = CompareLaps.alignLapPair(entryA, chanA.dStepM, entryB, chanB.dStepM)
        return Pair(
            aligned: aligned,
            laps: [
                Lap(id: 0, sessionId: rows[selA].sessionId, lapNum: rows[selA].lapNum, timeMs: entryA.timeMs),
                Lap(id: 1, sessionId: rows[selB].sessionId, lapNum: rows[selB].lapNum, timeMs: entryB.timeMs),
            ],
            metricsA: CompareLaps.lapMetrics(entryA),
            metricsB: CompareLaps.lapMetrics(entryB),
            mismatch: CompareLaps.lengthMismatchRatio(entryA, chanA.dStepM, entryB, chanB.dStepM)
        )
    }
}

File: apps/android/app/src/main/kotlin/app/trackevolution/screens/CompareLapsScreen.kt
package app.trackevolution.screens

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.PaddingValues
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.DropdownMenu
import androidx.compose.material3.DropdownMenuItem
import androidx.compose.material3.HorizontalDivider
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.unit.dp
import app.trackevolution.core.CompareLaps
import app.trackevolution.core.LapTime
import app.trackevolution.ui.LoadState
import app.trackevolution.ui.TEEmpty
import app.trackevolution.ui.TELoadable
import app.trackevolution.ui.charts.LapChannelChart
import app.trackevolution.ui.theme.TrackCard
import app.trackevolution.ui.theme.TrackTheme
import kotlin.math.abs
import kotlin.math.roundToInt

/**
 * Compare any two laps with telemetry at one track — different sessions, events,
 * even years apart (#165). `viewLapCompare` in `public/app.js` is the reference:
 * the same "current me vs best me" default picks, head-to-head numbers and
 * length-mismatch warning; the charts are the same [LapChannelChart] the event
 * page uses, fed the pair `CompareLaps.alignLapPair` builds.
 */
@Composable
fun CompareLapsScreen(
    model: CompareLapsModel,
    modifier: Modifier = Modifier,
) {
    val colors = TrackTheme.colors

    LaunchedEffect(Unit) { if (model.state == LoadState.Loading) model.load() }

    TELoadable(state = model.state, onRetry = model::load, modifier = modifier) {
        val pair = model.pair
        LazyColumn(
            modifier = Modifier.fillMaxSize().padding(horizontal = 16.dp),
            verticalArrangement = Arrangement.spacedBy(12.dp),
            contentPadding = PaddingValues(vertical = 12.dp),
        ) {
            item("head") {
                Text("Compare two laps", style = TrackTheme.typography.h1, color = colors.textStrong)
            }

            if (pair == null) {
                item("empty") {
                    TEEmpty(
                        "Comparing laps needs two laps with telemetry at this track — " +
                            "import a session (or record laps) first.",
                    )
                }
                return@LazyColumn
            }

            item("pickers") {
                Column(verticalArrangement = Arrangement.spacedBy(8.dp)) {
                    LapPicker("Lap A", colors.chartLine, model.selA, model, model::pickA)
                    LapPicker("Lap B", colors.chartLineB, model.selB, model, model::pickB)
                }
            }

            if (pair.mismatch > CompareLaps.LENGTH_MISMATCH_WARN) {
                item("mismatch") {
                    Text(
                        "⚠️ These laps cover driven distances ${(pair.mismatch * 100).roundToInt()}% apart — " +
                            "likely a different layout or start/finish line, so the distance alignment may be off.",
                        style = TrackTheme.typography.xs,
                        color = colors.textMuted,
                    )
                }
            }

            item("head-to-head") { HeadToHead(pair) }

            item("charts-hint") {
                Text(
                    "The delta chart shows where time is gained or lost vs the faster lap; " +
                        "the channels below show why.",
                    style = TrackTheme.typography.xs,
                    color = colors.textFaint,
                )
            }

            item("charts") {
                LapChannelChart(
                    channels = pair.aligned,
                    laps = pair.laps,
                    initialSelection = listOf(0, 1),
                )
            }
        }
    }
}

@Composable
private fun LapPicker(
    label: String,
    color: Color,
    selected: Int,
    model: CompareLapsModel,
    onPick: (Int) -> Unit,
) {
    val colors = TrackTheme.colors
    var expanded by remember { mutableStateOf(false) }

    Row(verticalAlignment = Alignment.CenterVertically, horizontalArrangement = Arrangement.spacedBy(8.dp)) {
        Box(Modifier.size(10.dp).background(color, CircleShape))
        Box {
            Text(
                model.pickLabel(selected),
                style = TrackTheme.typography.bodyStrong,
                color = colors.textStrong,
                maxLines = 1,
                modifier = Modifier
                    .background(colors.surfaceCard, RoundedCornerShape(TrackTheme.radii.sm))
                    .border(1.dp, colors.borderHairline, RoundedCornerShape(TrackTheme.radii.sm))
                    .clickable { expanded = true }
                    .padding(horizontal = 12.dp, vertical = 8.dp)
                    .semantics { contentDescription = "$label: ${model.pickLabel(selected)}" },
            )
            DropdownMenu(expanded = expanded, onDismissRequest = { expanded = false }) {
                model.rows.indices.forEach { index ->
                    DropdownMenuItem(
                        text = {
                            Text(
                                (if (index == selected) "✓ " else "") + model.pickLabel(index),
                                style = TrackTheme.typography.sm,
                            )
                        },
                        onClick = {
                            expanded = false
                            onPick(index)
                        },
                    )
                }
            }
        }
    }
}

private const val KPH_TO_MPH = 0.621371

@Composable
private fun HeadToHead(pair: CompareLapsModel.LapPair) {
    val mph = { kph: Double -> "${(kph * KPH_TO_MPH).roundToInt()} mph" }
    val mphDelta = { kph: Double -> signed(kph, "${(abs(kph) * KPH_TO_MPH).roundToInt()} mph") }
    val ppDelta = { d: Double -> signed(d, "${(abs(d) * 10).roundToInt() / 10.0}pp") }

    TrackCard(Modifier.fillMaxWidth()) {
        StatRow(
            "Lap time",
            LapTime.fmtMs(pair.metricsA.timeMs),
            LapTime.fmtMs(pair.metricsB.timeMs),
            LapTime.fmtDelta(pair.metricsB.timeMs - pair.metricsA.timeMs),
        )
        MetricRow("Top speed", pair.metricsA.topSpeedKph, pair.metricsB.topSpeedKph, mph, mphDelta)
        MetricRow("Min speed", pair.metricsA.minSpeedKph, pair.metricsB.minSpeedKph, mph, mphDelta)
        MetricRow("Avg speed", pair.metricsA.avgSpeedKph, pair.metricsB.avgSpeedKph, mph, mphDelta)
        MetricRow(
            "Max RPM", pair.metricsA.maxRpm, pair.metricsB.maxRpm,
            { "${it.roundToInt()}" }, { signed(it, "${abs(it).roundToInt()}") },
        )
        MetricRow(
            "Max lateral G", pair.metricsA.maxLatG, pair.metricsB.maxLatG,
            { String.format("%.2f", it) }, { signed(it, String.format("%.2f", abs(it))) },
        )
        MetricRow(
            "Full throttle", pair.metricsA.fullThrottlePct, pair.metricsB.fullThrottlePct,
            { "${it.roundToInt()}% of lap" }, ppDelta,
        )
        MetricRow(
            "On the brakes", pair.metricsA.brakingPct, pair.metricsB.brakingPct,
            { "${it.roundToInt()}% of lap" }, ppDelta,
            last = true,
        )
    }
}

@Composable
private fun MetricRow(
    label: String,
    a: Double?,
    b: Double?,
    fmt: (Double) -> String,
    deltaFmt: (Double) -> String,
    last: Boolean = false,
) {
    StatRow(
        label = label,
        a = a?.let(fmt) ?: "—",
        b = b?.let(fmt) ?: "—",
        delta = if (a != null && b != null) deltaFmt(b - a) else "—",
        last = last,
    )
}

@Composable
private fun StatRow(label: String, a: String, b: String, delta: String, last: Boolean = false) {
    val colors = TrackTheme.colors
    Row(
        modifier = Modifier
            .fillMaxWidth()
            .padding(vertical = 6.dp)
            .semantics { contentDescription = "$label: lap A $a, lap B $b, delta $delta" },
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(8.dp),
    ) {
        Text(label, style = TrackTheme.typography.sm, color = colors.textMuted, modifier = Modifier.weight(1f))
        Text(a, style = TrackTheme.typography.lapTime, color = colors.textStrong)
        Text(b, style = TrackTheme.typography.lapTime, color = colors.textStrong)
        Text(delta, style = TrackTheme.typography.lapTime, color = colors.textMuted)
    }
    if (!last) HorizontalDivider(color = colors.borderHairline)
}

/** The sign is the message, so it is always shown. */
private fun signed(value: Double, magnitude: String): String =
    if (value < 0) "−$magnitude" else "+$magnitude"
